File: python/2020/july/binary-tree-zigzag-order-traversal.py
from typing import List

from common import TreeNode


class Solution:
    def zig_zag(self, root: TreeNode) -> List[List[int]]:
        if not root: return []
        res, tmp, stack, flag = [], [], [root], 1
        while stack:
            for i in range(len(stack)):
                cur = stack.pop(0)
                tmp += [cur.val]
                if cur.left: stack += [cur.left]
                if cur.right: stack += [cur.right]
            res += tmp[::flag]
            tmp = []
            flag *= -1
        return res

File: code/go/2020/july/binary-tree-zigzag-order-traversal.go
package july

import (
	. "leetcode/common"
)

func bfs(s *[][]int, level int, root *TreeNode) {
	if root == nil {
		return
	}

	if len(*s) == level {
		*s = append(*s, []int{})
	}

	if level%2 == 1 {
		(*s)[level] = append([]int{root.Val}, (*s)[level]...)
	} else {
		(*s)[level] = append((*s)[level], root.Val)
	}
	for _, v := range []*TreeNode{root.Left, root.Right} {
		bfs(s, level+1, v)
	}
}

func zigzagLevelOrder(root *TreeNode) [][]int {
	if root == nil {
		return [][]int{}
	}
	var s [][]int
	bfs(&s, 0, root)
	return s
}
